File: str/replace/stringReplace.py
# replace with longer or shorter
def replaceStr(input, source, target):
    if len(source) < len(target):
        return replace_longer(list(input), source, target)
    else:
        return replace_shorter(list(input), source, target)
# slow, fast
def replace_shorter(arr, source, target):
    slow = 0
    fast = 0
    while fast < len(arr):
        # need also consider space that does not inlucde source substring
        if fast <= len(arr) - len(source) and equalsubstr(arr, fast, source):
            replacestr(arr, slow, target)
            slow += len(target)
            fast += len(source)
        else:
            arr[slow] = arr[fast]
            slow += 1
            fast += 1
    return ''.join(arr[:slow])

def equalSubStr(arr, fast, source):
    temp = list(source)
    for index in range(len(source)):
        if arr[index + fast] != temp[index]:
            return False
    return True

def replacestr(arr, slow, target):
    temp = list(target)
    for index in range(len(target)):
        arr[index + slow] = temp[index]
# slow, fast, last_index
def replace_longer(arr, source, target):
    target_index_list = get_target_end_index(arr, target)
    writing_index = len(arr) + len(target_index) * (len(target) - len(source)) - 1  # writing index
    traverse_index = len(arr) - 1
    last_index = len(target_index_list) - 1

    while traverse_index >= 0:
        if(last_index >= 0 and traverse_index == target_index_list[last_index]):
            copySubStr(arr, writing_index - len(target) + 1, target)
            writing_index -= len(target)
            traverse_index -= len(source)
            last_index -= 1
        else:
            arr[writing_index] = arr[traverse_index]
            writing_index -= 1
            traverse_index -=1
    return ''.join(arr)

# record end index
def get_target_end_index(arr, source, target):
    res = []
    start_index = 0
    while start_index <= len(arr) - len(source):
        if equalSubStr(arr, start_index, source):
            res.append(start_index + len(source) - 1)   # record end index
            start_index += len(source)
        else:
            start_index += 1
    return res

File: str/replace/stringReplace.java
//Q6: string replace
//Given an original string input, and two strings S and T, replace all occurrences of S in input with T.
//solution1: not using StringBuilder
//assuming inputs are not null
public String replace(String input, String s, String t) {
    char[] arr = input.toCharArray();
    if(s.length() >= t.length())
    	return replaceShorter(arr, s, t);
    else
    	return replaceLonger(arr, s, t);
}
private String replaceShorter(char[] arr, String s, String t){
	int slow = 0;  //tracking resulting string , slow not included
	int fast = 0;  //traverse arr
  //make sure we going outside after finish the last replacement
	while(fast < arr.length){
    //fast index goes to the last starting position of s && need to replace
		if(fast <= arr.length - s.length() && equalSubStr(arr, fast, s)){
			copySubStr(arr, slow, t);
			slow += t.length();
			fast += s.length();
    //no need to replace
		}else{
			arr[slow++] = arr[fast++];
		}
	}
	return new String(arr, 0, slow);
}
private String replaceLonger(char[] arr, String s, String t){
    //get all ending index match s in arr
    List<Integer> matches = getAllMathces(arr, s);
  	char[] res = new char[arr.length + matches.size() * (t.length() - s.length())];
  	int lastIndex = matches.size()-1;
    int fast = arr.length -1;   //traverse arr
    int slow = res.length -1;   //tracking resulting string

    while(fast >= 0){
    	if(lastIndex >= 0 && fast == matches.get(lastIndex)){
      	copySubStr(res, slow - t.length() + 1, t);
        slow -= t.length();
        fast -= s.length();
        lastIndex--;
      }else{
      	res[slow--] = arr[fast--];
      }
    }
    return new String(res);
  }
  private List<Integer> getAllMathces(char[] arr, String s){
    //all ending matching index of s in arr
  	List<Integer> matches = new ArrayList<>();
    int start_index = 0;
    //for all possible starting index
    while(start_index <= arr.length - s.length()){
    	if(equalSubStr(arr, start_index, s)){
      	matches.add(start_index + s.length() - 1);  //record the end index
        start_index += s.length(); //check next possible starting index
      }else{
      	start_index++;
      }
    }
    return matches;
  }

  private boolean equalSubStr(char[] arr, int fromIndex, String s){
    //char[] ss = s.toCharArray();
  	for(int i = 0; i < s.length(); i++){
    	if(arr[i+fromIndex] != s.charAt(i))
        return false;
    }
    return true;
  }
  private void copySubStr(char[] arr, int fromIndex, String t){
  	for(int i = 0; i < t.length(); i++){
    	arr[i+fromIndex] = t.charAt(i);
    }
  }
